Objects/bytearrayobject.c

```diff
@@ -2900,27 +2913,39 @@ bytearrayiter_reduce(PyObject *self, PyObject *Py_UNUSED(ignored))
      * call must be before access of iterator pointers.
      * see issue #101765 */
     bytesiterobject *it = _bytesiterobject_CAST(self);
-    if (it->it_seq != NULL) {
-        return Py_BuildValue("N(O)n", iter, it->it_seq, it->it_index);
-    } else {
-        return Py_BuildValue("N(())", iter);
+    PyObject *ret = NULL;
+    Py_ssize_t index = FT_ATOMIC_LOAD_SSIZE_RELAXED(it->it_index);
+    if (index >= 0) {
+        if (index <= PyByteArray_GET_SIZE(it->it_seq)) {
+            ret = Py_BuildValue("N(O)n", iter, it->it_seq, index);
+        }
     }
+    if (ret == NULL) {
+        ret = Py_BuildValue("N(())", iter);
+    }
+    return ret;
 }
 
 static PyObject *
 bytearrayiter_setstate(PyObject *self, PyObject *state)
 {
     Py_ssize_t index = PyLong_AsSsize_t(state);
-    if (index == -1 && PyErr_Occurred())
+    if (index == -1 && PyErr_Occurred()) {
         return NULL;
+    }
 
     bytesiterobject *it = _bytesiterobject_CAST(self);
-    if (it->it_seq != NULL) {
-        if (index < 0)
-            index = 0;
-        else if (index > PyByteArray_GET_SIZE(it->it_seq))
-            index = PyByteArray_GET_SIZE(it->it_seq); /* iterator exhausted */
-        it->it_index = index;
+    if (FT_ATOMIC_LOAD_SSIZE_RELAXED(it->it_index) >= 0) {
+        if (index < -1) {
```

Comment:
Why is it changed to `index < -1` rather than previously `index < 0`?

Reply:
`-1` indicates exhaustion so comparing `< 0` or `< -1` as floor doesn't make a difference. `list` compares with `-1` so was just being consistent.

Reply:
I found some thread safety in list iterator code, filed https://github.com/python/cpython/issues/130263